### PR TITLE
Feature ETP-3903: expand printdocumentws tests

### PR DESCRIPTION
**Pull request by RomanMagnoli**
This pull request significantly expands and improves the test coverage for the `PrintDocument` class, focusing on scenarios involving custom and base document handling, error reporting, and delegation logic for report managers. It introduces new test cases to verify the correct behavior of multi-document printing, error messages, and custom template handling. Additionally, the changes update imports to use Jakarta EE and add helpful test utilities.

Key changes include:

### Expanded Test Coverage for Document Handling

* Added tests to verify that `doGet` correctly generates both base and custom documents, prints multi-reports, and handles scenarios where only base documents exist. These tests ensure the correct invocation of document generation and report printing methods, and validate the contents of `customDocuments` and document type assignment.
* Added a test that checks error messages include custom document numbers when certain custom documents are missing, improving the reliability and clarity of error reporting.

### Improved Report Manager Delegation Logic

* Updated tests for `getReportManager` to verify that it delegates to the existing report manager instance when appropriate (e.g., when not custom, when template location is empty, or when there are no active templates), using `assertSame` instead of `assertNull` and setting up the `reportManager` field. [[1]](diffhunk://#diff-f9802848179bdb5c9f68c5852ad56009385fcf9e9dbd0b380a53bbad64107bb7R306-R311) [[2]](diffhunk://#diff-f9802848179bdb5c9f68c5852ad56009385fcf9e9dbd0b380a53bbad64107bb7R322-R330) [[3]](diffhunk://#diff-f9802848179bdb5c9f68c5852ad56009385fcf9e9dbd0b380a53bbad64107bb7R341-R348)
* Added a test to ensure that an exception is thrown when a custom template requires servlet configuration, increasing robustness for custom template scenarios.

### Test Infrastructure and Import Updates

* Updated imports to use Jakarta EE (`jakarta.servlet` and related classes), added new mock fields for `HttpSession` and `ServletContext`, and included utility methods for setting parent fields and adding saved reports in tests. [[1]](diffhunk://#diff-f9802848179bdb5c9f68c5852ad56009385fcf9e9dbd0b380a53bbad64107bb7L21-R66) [[2]](diffhunk://#diff-f9802848179bdb5c9f68c5852ad56009385fcf9e9dbd0b380a53bbad64107bb7R87-R92) [[3]](diffhunk://#diff-f9802848179bdb5c9f68c5852ad56009385fcf9e9dbd0b380a53bbad64107bb7R360-R496)

These changes collectively enhance the reliability and maintainability of the `PrintDocument` component by ensuring comprehensive test coverage for both standard and edge-case behaviors.

### CODE REVIEW

File: src-test/src/com/etendoerp/printdocumentws/PrintDocumentTest.java [hunk 1]
```diff
@@ -18,38 +18,52 @@
...
 
+import org.openbravo.base.exception.OBException;
+
+
 /**
  * Tests for {@link PrintDocument}.
  */
 @ExtendWith(MockitoExtension.class)
+@MockitoSettings(strictness = Strictness.LENIENT)
 class PrintDocumentTest {
```
Comment (etendobot) [inline src-test/src/com/etendoerp/printdocumentws/PrintDocumentTest.java:64]: **Suggestion**

Consider if strictness can be maintained at the default level by moving shared stubs to individual test methods or using @BeforeEach selectively.

```java
// Prefer using default strictness and stub only what is needed for each test case.
```

**Rationale**: Lenient strictness allows mocks to be stubbed without being called, which can lead to brittle tests and hide unnecessary configuration. While sometimes necessary for complex setups, it's a best practice to aim for strict stubbing.

File: src-test/src/com/etendoerp/printdocumentws/PrintDocumentTest.java [hunk 2]
```diff
@@ -329,9 +357,141 @@ void testCustomDocumentsInitiallyEmpty() {
...
+    DocumentTemplate mockTemplate = mock(DocumentTemplate.class);
+
+    when(mockTemplate.isActive()).thenReturn(true);
+    when(mockTemplate.getTemplateLocation()).thenReturn("@basedesign@/custom.jrxml");
+    when(mockDocType.getDocumentTemplateList()).thenReturn(Collections.singletonList(mockTemplate));
+    when(mockRequest.getSession()).thenReturn(mockSession);
+    when(mockSession.getServletContext()).thenReturn(mockServletContext);
+
+    assertThrows(Exception.class, () -> printDocument.getReportManager(mockRequest, mockDocType, true));
+  }
```
Comment (etendobot) [inline src-test/src/com/etendoerp/printdocumentws/PrintDocumentTest.java:468]: **Suggestion**

Identify the specific exception type expected to be thrown (e.g., ServletException or a custom Etendo exception) and use it in the assertion.

```java
assertThrows(SpecificExpectedException.class, () -> printDocument.getReportManager(mockRequest, mockDocType, true));
```

**Rationale**: Catching a generic Exception.class in a test can be misleading, as the test might pass due to a different exception than the one intended (e.g., a NullPointerException). Specifying the exact exception type makes the test intent clearer and more reliable.

**Review by etendobot — APPROVED** [review]

**Review by etendobot — APPROVED** [review]

File: src-test/src/com/etendoerp/printdocumentws/PrintDocumentTest.java [hunk 3]
```diff
@@ -18,38 +18,49 @@
...
+import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
 import java.lang.reflect.Field;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 
+import javax.servlet.ServletContext;
 import javax.servlet.http.HttpServletRequest;
```
Comment (etendobot) [inline src-test/src/com/etendoerp/printdocumentws/PrintDocumentTest.java:40]: **Suggestion**

The switch from jakarta.servlet to javax.servlet is correct for the current Etendo environment. Ensure all other web-related components in the module follow this pattern.

```java
import javax.servlet.ServletContext;
```

**Rationale**: Etendo currently targets environments using Tomcat 9, which implements the J2EE/Jakarta EE 8 specifications using the javax.servlet package. Aligning with this package is necessary for compatibility with the core framework.

File: src-test/src/com/etendoerp/printdocumentws/PrintDocumentTest.java [hunk 4]
```diff
@@ -329,9 +358,148 @@ void testCustomDocumentsInitiallyEmpty() {
...
+
+  private static void setParentField(Object target, Object value) {
+    try {
+      Field field = com.smf.ws.printdocument.PrintDocument.class.getDeclaredField(
+          PrintDocumentTest.REPORT_MANAGER_FIELD);
+      field.setAccessible(true);
+      field.set(target, value);
+    } catch (ReflectiveOperationException e) {
+      throw new TestSetupException("Unable to set parent field: " + PrintDocumentTest.REPORT_MANAGER_FIELD, e);
+    }
```
Comment (etendobot) [inline src-test/src/com/etendoerp/printdocumentws/PrintDocumentTest.java:484]: **Suggestion**

Remove the 'PrintDocumentTest.' prefix when accessing the REPORT_MANAGER_FIELD constant within the same class.

```java
throw new TestSetupException("Unable to set parent field: " + REPORT_MANAGER_FIELD, e);
```

**Rationale**: Inside the PrintDocumentTest class, referencing its own static constants with the class name prefix is redundant and can be simplified for better readability.

File: src-test/src/com/etendoerp/printdocumentws/ReportManagerCustomTest.java [hunk 5]
```diff
@@ -386,12 +389,12 @@ private String getPrivateField(ReportManagerCustom instance, String fieldName) t
...
-   * @param fieldName the name of the private field
    * @return the field value
-   * @throws Exception if reflective access fails
+   * @throws Exception
+   *     if reflective access fails
    */
-  private Object getPrivateFieldObject(String fieldName) throws Exception {
-    Field field = ReportManagerCustom.class.getDeclaredField(fieldName);
+  private Object getPrivateFieldObject() throws Exception {
+    Field field = ReportManagerCustom.class.getDeclaredField("_connectionProvider");
```
Comment (etendobot) [inline src-test/src/com/etendoerp/printdocumentws/ReportManagerCustomTest.java]: **Suggestion**

Maintain the 'fieldName' parameter in the 'getPrivateFieldObject' helper method to allow retrieving different private fields.

```java
private Object getPrivateFieldObject(String fieldName) throws Exception {
```

**Rationale**: Removing the 'fieldName' parameter and hardcoding the field name within the helper method reduces its flexibility and reusability for other test scenarios.

File: src-test/src/com/etendoerp/printdocumentws/ReportManagerCustomTest.java [hunk 6]
```diff
@@ -386,12 +389,12 @@ private String getPrivateField(ReportManagerCustom instance, String fieldName) t
...
     field.setAccessible(true);
```
Comment (etendobot) [inline src-test/src/com/etendoerp/printdocumentws/ReportManagerCustomTest.java]: **Suggestion**

Use a parameter for the field name instead of hardcoding '_connectionProvider'.

```java
Field field = ReportManagerCustom.class.getDeclaredField(fieldName);
```

**Rationale**: Hardcoding field names in reflection-based helpers makes the test suite harder to maintain and less adaptable to changes in the class structure.

File: src-test/src/com/etendoerp/printdocumentws/ReportManagerCustomTest.java [hunk 7]
```diff
@@ -302,21 +302,24 @@ void testProcessReportThrowsReportingExceptionOnError() {
...
+   * @param templateInfo
+   *     the template info to set on the mock report, may be {@code null}
    * @return the populated design parameters map
-   * @throws Exception if reflective invocation fails
+   * @throws Exception
+   *     if reflective invocation fails
    */
-  private HashMap<String, Object> invokePopulateWithDefaults(String documentId, String language,
+  private HashMap<String, Object> invokePopulateWithDefaults(String language,
       TemplateInfo templateInfo) throws Exception {
```
Comment (etendobot) [inline src-test/src/com/etendoerp/printdocumentws/ReportManagerCustomTest.java]: **Suggestion**

Restore the 'documentId' parameter to 'invokePopulateWithDefaults'.

```java
private HashMap<String, Object> invokePopulateWithDefaults(String documentId, String language,
```

**Rationale**: The helper method 'invokePopulateWithDefaults' lost its ability to test different document IDs, which obscures the specific data used in each test case.

File: src-test/src/com/etendoerp/printdocumentws/ReportManagerCustomTest.java [hunk 8]
```diff
@@ -302,21 +302,24 @@ void testProcessReportThrowsReportingExceptionOnError() {
...
     Report mockReport = mock(Report.class);
     VariablesSecureApp mockVars = createMockVariables(language);
-    when(mockReport.getDocumentId()).thenReturn(documentId);
+    when(mockReport.getDocumentId()).thenReturn(ReportManagerCustomTest.DOC_123);
     when(mockReport.getTemplateInfo()).thenReturn(templateInfo);
```
Comment (etendobot) [inline src-test/src/com/etendoerp/printdocumentws/ReportManagerCustomTest.java:318]: **Suggestion**

Use the method parameter for the document ID return value instead of a class constant.

```java
when(mockReport.getDocumentId()).thenReturn(documentId);
```

**Rationale**: Hardcoding the document ID value inside a shared helper method makes individual tests less explicit about the data they are validating.

File: src-test/src/com/etendoerp/printdocumentws/ReportManagerCustomTest.java [hunk 9]
```diff
@@ -302,21 +302,24 @@ void testProcessReportThrowsReportingExceptionOnError() {
...
+
   /**
    * Creates mock {@link Report} and {@link VariablesSecureApp} objects and invokes
    * the private {@code populateDesignParameters} method with the given arguments.
    *
-   * @param documentId the document identifier to set on the mock report
-   * @param language the language code to set on the mock variables
-   * @param templateInfo the template info to set on the mock report, may be {@code null}
+   * @param language
+   *     the language code to set on the mock variables
```
Comment (etendobot) [inline src-test/src/com/etendoerp/printdocumentws/ReportManagerCustomTest.java:308]: **Suggestion**

Use single-line documentation for Javadoc parameters.

```java
* @param language the language code to set on the mock variables
```

**Rationale**: The new multiline formatting for Javadoc parameters is inconsistent with the standard Etendo documentation style.

**Review by etendobot — APPROVED** [review]

**Review by etendobot — APPROVED** [review]

**Review by etendobot — APPROVED** [review]

**Review by etendobot — APPROVED** [review]

File: build.gradle [hunk 10]
```diff
@@ -48,13 +48,24 @@ repositories {
     }
 }
 
+sourceSets {
+    main {
+        resources {
+            srcDirs("etendo-resources")
+        }
```
Comment (etendobot) [inline build.gradle:55]: > [!WARNING]
**⚠️ Blocking Issue**

Restore the java source directory mapping in the main sourceSet.

```gradle
        java {
            srcDirs("src")
        }
        resources {
            srcDirs("etendo-resources")
        }
```

**Rationale**: The removal of the java { srcDirs('src') } block in the main sourceSet will prevent Gradle from finding the module's Java source files if they follow the standard Etendo structure. This will lead to compilation failures.

File: build.gradle [hunk 11]
```diff
@@ -48,13 +48,24 @@ repositories {
...
+}
+
 /**
 * Declare Java dependencies using 'implementation'
 * Ex: implementation "com.sun.mail:javax.mail:1.6.2"
 */
 dependencies {
-
    implementation('com.smf:ws.printdocument:3.0.0')
    implementation('com.etendoerp.platform:etendo-core:[26.1.0,26.2.0)')
```
Comment (etendobot) [inline build.gradle:65]: > [!WARNING]
**⚠️ Blocking Issue**

Restore the test dependencies (JUnit, Mockito) and the conditional implementation logic for local developments.

```gradle
   if (findProject(':modules:com.smf.ws.printdocument') != null) {
      implementation(project(':modules:com.smf.ws.printdocument'))
   } else {
      implementation('com.smf:ws.printdocument:3.0.0')
   }

    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.8.1'
    testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.8.1'
    testImplementation 'org.mockito:mockito-junit-jupiter:5.12.0'
```

**Rationale**: The PR objective is to 'expand tests', but the diff removes the test dependencies and the conditional logic for local module development. Without these dependencies and the test sourceSet (removed in the same block), it is impossible to run or expand tests.

File: build.gradle [hunk 12]
```diff
@@ -48,13 +48,24 @@ repositories {
...
```
Comment (etendobot) [inline build.gradle:66]: > [!WARNING]
**⚠️ Blocking Issue**

Use a valid Etendo Core version or restore the standard coreCompileClasspath reference if this module is built as part of an Etendo environment.

```gradle
   compileOnly files(rootProject.sourceSets.main.output, rootProject.configurations.compileClasspath.filter { !it.path.contains('/modules/') })
```

**Rationale**: The core version range [26.1.0, 26.2.0) refers to a version that does not currently exist (latest stable is in the 24.x range). Additionally, modules typically depend on core via the root project's classpath in source environments to ensure compatibility and avoid duplication.

**Review by etendobot — DISMISSED** [review]
Changes requested by agent. Please resolve blocking issues.

**Review by isaiasb-etendo — APPROVED** [review]
